### ISSUE TITLE
Missing bits of statinfo accessors in FilePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4712]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#4712|
|Type|enhancement|
|Created|2010-10-25 15:42:21Z|
|Branch|https://github.com/twisted/twisted/tree/provide-statinfo-accessors-4712|

FilePath currently provides the following methods to get information in statinfo:

 * `getsize()`     (statinfo.st_size)
 * `getModificationTime()`    (statinfo.st_mtime)
 * `getStatusChangeTime()`     (statinfo.st_ctime)
 * `getAccessTime()`    (statinfo.st_atime)
 * `exists()`    (whether statinfo even exists)
 * `isdir()`    (part of st.st_mode)
 * `isfile()`    (part of st.st_mode)
 * `islink()`    (part of st.st_mode)

The following statinfo keys have no accessors  ([meanings source](http://www.gnu.org/s/libc/manual/html_node/Attribute-Meanings.html)):

 * `st_ino` (The file serial number, which distinguishes this file from all other files on the same device)
 * `st_dev` (Identifies the device containing the file. The st_ino and st_dev, taken together, uniquely identify the file. The st_dev value is not necessarily consistent across reboots or system crashes, however.)
 * `st_nlink` (The number of hard links to the file. This count keeps track of how many directories have entries for this file. If the count is ever decremented to zero, then the file itself is discarded as soon as no process still holds it open. Symbolic links are not counted in the total.)
 * `st_uid` (The user ID of the file's owner.)
 * `st_gid` (The group ID of the file)

If `FilePath.statinfo` is going to be deprecated, accessors to its keys should be provided.
Perhaps some of these (such as `st_ino` and `st_dev`) should not be exposed, but `twisted.protocols.ftp` for instance references `st_nlink`, `st_uid`, and `st_gid`; and `twisted.web2.static` references `st_ino`.

Note that this ticket is holding up [#4711](https://github.com/twisted/twisted/issues/4711) and possibly [#4450](https://github.com/twisted/twisted/issues/4450)

<details><summary>Searchable metadata</summary>

```
trac-id__4712 4712
type__enhancement enhancement
reporter__cyli cyli
priority__normal normal
milestone__ 
branch__branches_provide_statinfo_accessors_4712 branches/provide-statinfo-accessors-4712
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1288021341000000 1288021341000000
changetime__1295222425000000 1295222425000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [30174]) Branching to 'provide-statinfo-accessors-4712'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [30408]) Added access methods to get FilePath.statinfo's st_ino, st_dev, st_nlink, st_uid, and st_gid fields

refs [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

(In [30409]) Ensure that the device number is a long.

refs [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [30410]) Ensure that the Inode Number is also a long.

refs [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

Buildbot results:  http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/provide-statinfo-accessors-4712

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set owner to @cyli|
|-|-|

All new methods need an `@since` marker.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

(In [30411]) Added a @since to all the methods

refs [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @cyli|
|-|-|

Hooray

  1. in `test_statinfBitsAreNumbers`, any reason not to use `assertIsInstance`?  Not quite the same check, since it will allow subclasses, but is that really the concern?  Aside from that, `assertEquals` is preferred over `failUnlessEqual`.
  1. It might also be better, in `test_statinfBitsAreNumbers`, to populate the `FilePath` instance with a known `stat_result` object and make sure the right values come back from the right method calls.  `return 0` would appear to be a valid implementation at the moment.  This would also allow testing of the `restat` logic, which is currently uncovered (by returning the known `stat_result` from a monkey-patched `restat` method on the instance, for example).
  1. Copyright dates should be bumped to include 2011.
  1. The news file should go in `twisted/topfiles/` instead of `twisted/news/topfiles` :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

(In [30467]) Added a new test to monkey-patch restat, and ensure accessors are returning the correct values.

refs [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli commented|
|-|-|

>   1. in `test_statinfBitsAreNumbers`, any reason not to use `assertIsInstance`?  Not quite the same check, since it will allow subclasses, but is that really the concern?  Aside from that, `assertEquals` is preferred over `failUnlessEqual`.

Mainly laziness, so I don't have to remember what is implemented as what (e.g. isinstance(True, int) is True).  But in this case, int and long are separate, so using assertIsInstance wouldn't be a problem.  (Also realized I made a typo in the method name)  Changed.

>   1. It might also be better, in `test_statinfBitsAreNumbers`, to populate the `FilePath` instance with a known `stat_result` object and make sure the right values come back from the right method calls.  `return 0` would appear to be a valid implementation at the moment.  This would also allow testing of the `restat` logic, which is currently uncovered (by returning the known `stat_result` from a monkey-patched `restat` method on the instance, for example).

I put this part in a different test, since I was trying to ensure in the first test that the values returns were of the valid type.

>   1. Copyright dates should be bumped to include 2011.

Done

>   1. The news file should go in `twisted/topfiles/` instead of `twisted/news/topfiles` :)

Done

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli removed owner|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/provide-statinfo-accessors-4712)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @cyli|
|-|-|

There's a stray print in `test_statinfoNumbersAreValid`.  Otherwise, looks great.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_17"></a>|@cyli commented|
|-|-|

(In [30495]) Remove stray print

refs [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_18"></a>|@cyli set status to `closed`|
|-|-|

(In [30497]) Merge provide-statinfo-accessors-4712: accessors to get the inode, device, hand hard link numbers, as well as uid and gid

Author: cyli
Reviewer: thijs, exarkun
Fixes: [#4712](https://github.com/twisted/twisted/issues/4712)

Provide accessors to get the inode number, device number, number of hard links, the user ID number, and group ID number of a FilePath file.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_19"></a>|@cyli set status to `reopened`|
|-|-|

(In [30500]) Reverse the merge because changing 'twisted' to 'Twisted' in Versions causes a failure in one of the new tests

reopens [#4712](https://github.com/twisted/twisted/issues/4712)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_20"></a>|@cyli set status to `closed`|
|-|-|

(In [30501]) Merge provide-statinfo-accessors-4712: accessors to get the inode, device, hand hard link numbers, as well as uid and gid

Author: cyli Reviewer: thijs, exarkun Fixes: [#4712](https://github.com/twisted/twisted/issues/4712)

Provide accessors to get the inode number, device number, number of hard links, the user ID number, and group ID number of a FilePath file.  (Re-opened wrong ticket)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
